### PR TITLE
Restrict some types in object initializer setters when not directly settable

### LINKED ISSUE #74331: Avoid recommending get-only properties without valid settable options in object initializer
## Version Used
VS 2022 Version 17.10.3
Roslyn 4.10.0 (04fb39164c99c)

Project info
C# 12.0
.NET 8.0

## Steps to Reproduce

Consider the following code:

```csharp
public sealed class FreightModel
{
    public double WeightKg { get; init; }
    
    public string Name { get; } = "some name";
}

public class ModelFactory
{
    public FreightModel Create(double weight)
    {
        return new()
        {
            // $$
        };
    }
}
```

In the above example code, place the cursor at `$$`, type `Na`, and notice the autocomplete result including the `Name` get-only property.

## Expected Behavior

We don't get the get-only string property

Reason: we don't need a string get-only property the object initializer. We cannot use that property in any way. The only scenario in which we want to show get-only properties is when they are set using an object or list initializer without the construction of a new instance. For example:
```csharp
public class Thing
{
    public List<string> SomeList { get; } = [];
}

new Thing()
{
    // we are only adding the string below to the list, without setting the property itself
    SomeList = 
    {
        "some string within the list"
    }
}
```

## Actual Behavior

The get-only property's name is suggested.

## Notes

Properties can be set in the following ways:
- Setting their value directly to an instance expression
- Setting their nested properties with an object initializer block
- Adding elements with a list initializer block

When none of the above are applicable, the name of the property should not be suggested.

### PR DESCRIPTION
**Pull request by Rekkonnect**
Fixes #74331 

## Implementation details

When the property or field is not directly settable, we do **not** recommend the member as settable if any of the following is met:
- Type matches the types `Object`, `String`, `Enum`, `Array`, `Delegate`, `MulticastDelegate`, `IEnumerable` and `IEnumerator`
- Type is a generic substitution of `IEnumerable<T>` or `IEnumerator<T>`
- Type is a delegate, struct, pointer or function pointer

The above types have no settable members within an object initializer and cannot make use of a collection initializer. The only legal syntax they can have is `X = { }` which is a no-op setter, and since it's more of a trick than is being actually used, we do not offer the user the recommendation. They are still free to use the unrecommended members in the initializer, but only being able to "assign" `{ }` as shown above.

We only filter `Array` itself because it does not support indexing. Array types with a concrete rank can be indexed within the initializer like so:
```csharp
new()
{
    // assume Array2D is int[,]
    Array2D = { [0, 0] = value, [0, 1] = value1 },
}
```

## Limitations

This is limited to just the known types as specified above. Other types that could be filtered out with the same properties, which would require traversing their members on-demand. This is a risky option because of the performance implications that could heavily affect the performance of the recommendation service which is intended to operate in real-time. It would be nice to have, but my personal view on solving that leans towards a cached value showing whether the type symbol can support object or collection initializers, being synchronized as the compilation is updated. Such a system would be very complicated and is not taken yet into consideration.

### CODE REVIEW

**Comment by Rekkonnect** [comment]
@CyrusNajmabadi ptal, let me know if the approach is appropriate

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 1]
```diff
@@ -119,13 +119,47 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
 
         throw ExceptionUtilities.Unreachable();
     }
+
+    private static bool MemberTypeCanSupportObjectInitializer(ITypeSymbol type)
+    {
+        switch (type.SpecialType)
+        {
+            // While it is perfectly legal to assign instances of all these types to `{}`,
+            // it has no effects and is thus a needless recommendation
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:137]: i don't understand this comment.

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 2]
```diff
@@ -119,13 +119,47 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
+                return false;
+        }
+
+        if (type is INamedTypeSymbol { IsGenericType: true } named)
+        {
+            var definition = named.OriginalDefinition;
+            switch (definition.SpecialType)
+            {
+                case SpecialType.System_Collections_Generic_IEnumerable_T:
+                case SpecialType.System_Collections_Generic_IEnumerator_T:
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:155]: these need explanations.

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 3]
```diff
@@ -119,13 +119,47 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
+                    return false;
+            }
+        }
+
+        return !type.IsDelegateType()
+            && !type.IsStructType()
+            && !type.IsFunctionPointerType()
+            && type.TypeKind != TypeKind.Pointer;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:163]: these need explanations.

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 1 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:137]: I mean assignments of the form `Property = {}` in the object initializer. Example of such a legal case:
```csharp
class C
{
    string S { get; }
}

new C()
{
    S = {},
};
```

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:165]: these need explanations.

Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:179]: Added explanations

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 3 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:190]: Those types have explanations provided above, ptal

**Comment by Rekkonnect** [comment]
@CyrusNajmabadi ready for review with resolved comments

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 4]
```diff
@@ -119,13 +119,74 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
+            case SpecialType.System_Enum:
+            case SpecialType.System_String:
+            case SpecialType.System_Object:
+            case SpecialType.System_Delegate:
+            case SpecialType.System_MulticastDelegate:
+
+            // We cannot use collection initializers in Array members,
+            // but for members of an array type with a specified rank we can
+            // For example, assuming Array2D is int[,]:
+            // Array2D = { [0, 0] = value, [0, 1] = value1 },
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:161]: I don't understand this comment at all. It seems like the comment is saying you can do this, but then we're still filtering out the member by returning false?

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 5]
```diff
@@ -119,13 +119,74 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
+            case SpecialType.System_Array:
+
+            // We cannot add to an enumerable or enumerator
+            // so we cannot use a collection initializer
+            case SpecialType.System_Collections_IEnumerable:
+            case SpecialType.System_Collections_IEnumerator:
+                return false;
+        }
+
+        if (type is INamedTypeSymbol { IsGenericType: true } named)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:171]: Can remove this if check

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs [hunk 6]
```diff
@@ -119,13 +119,74 @@ private static bool CanSupportObjectInitializer(ISymbol symbol)
...
+        {
+            var definition = named.OriginalDefinition;
+            switch (definition.SpecialType)
+            {
+                // We cannot add to an enumerable or enumerator
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:176]: These cases can merge into the upper switch. Just always do the original definition unwrapping

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 4 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:161]: If Array2D was `Array`, like so: `Array Array2D = new int[2, 2];`, then we would not be able to use a collection initializer. I should say "symbols of type Array" to further clarify that.

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 6 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs]: Applied changes

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 5 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs]: Applied changes

File: src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs (hunk 4 above)
Comment (Rekkonnect) [reply src/Features/Core/Portable/Completion/Providers/AbstractObjectInitializerCompletionProvider.cs:163]: Applied changes to the comment

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**